sync_module/selective_sync/selective_sync.go

```diff
@@ -0,0 +1,167 @@
+
+	"github.com/edaniels/golog"
+	"github.com/pkg/errors"
+
+	"go.viam.com/rdk/components/camera"
+	"go.viam.com/rdk/components/generic"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/services/datamanager"
+	"go.viam.com/rdk/services/vision"
+	"go.viam.com/utils"
+)
+
+var (
+	// Model is the full model definition.
+	Model = resource.NewModel("selectivesync", "demo", "vision")
+)
+
+func init() {
+	registration := resource.Registration[resource.Resource, *Config]{
+		Constructor: func(ctx context.Context, deps resource.Dependencies, conf resource.Config, logger golog.Logger) (resource.Resource, error) {
+			return newSelectiveSyncer(ctx, deps, conf, logger)
+		},
+	}
+	resource.RegisterComponent(generic.API, Model, registration)
+
+}
+
+// Config contains the name to the underlying camera and the name of the vision service to be used.
+type Config struct {
+	DataManager   string `json:"data_manager"`
+	Camera        string `json:"camera"`
+	VisionService string `json:"vision_service"`
+}
+
+// Validate validates the config and returns implicit dependencies.
+func (cfg *Config) Validate(path string) ([]string, error) {
+	if cfg.DataManager == "" {
+		return nil, fmt.Errorf(`expected "data_manager" attribute in %q`, path)
+	}
+	if cfg.Camera == "" {
+		return nil, fmt.Errorf(`expected "camera" attribute in %q`, path)
+	}
+	if cfg.VisionService == "" {
+		return nil, fmt.Errorf(`expected "vision_service" attribute in %q`, path)
+	}
+
+	return []string{cfg.DataManager, cfg.Camera, cfg.VisionService}, nil
+}
+
+type visionSyncer struct {
+	resource.Named
+	actualDM      datamanager.Service
+	camera        camera.Camera
+	visionService vision.Service
+
+	cancelCtx  context.Context
+	cancelFunc func()
+	logger     golog.Logger
+}
+
+func newSelectiveSyncer(ctx context.Context, deps resource.Dependencies, conf resource.Config, logger golog.Logger) (resource.Resource, error) {
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+	v := &visionSyncer{
+		Named:      conf.ResourceName().AsNamed(),
+		logger:     logger,
+		cancelCtx:  cancelCtx,
+		cancelFunc: cancelFunc,
+	}
+	if err := v.Reconfigure(ctx, deps, conf); err != nil {
+		return nil, err
+	}
+	v.startBackgroundProcess()
+	return v, nil
+}
+
+func (s *visionSyncer) Reconfigure(ctx context.Context, deps resource.Dependencies, conf resource.Config) error {
+	dmConfig, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return err
+	}
+
+	s.actualDM, err = datamanager.FromDependencies(deps, dmConfig.DataManager)
+	if err != nil {
+		return errors.Wrapf(err, "unable to get dm %v for visionSyncer", dmConfig.DataManager)
+	}
+	// If sync_disabled is false for the data manager, there are no guarantees that the selective sync
+	// behavior will function as intended. It is possible that data will be synced irregardless of the trigger.
+	s.logger.Warnw("Behavior is undetermined when scheduled sync is enabled and using selective sync", "data manager", dmConfig.DataManager)
```

Comment:
Originally, I thought we might want sync to be disabled for this, but now I am not so convinced. I think it's perfectly valid for there to be a scheduled sync and selective trigger that may occur before that sync time. I also think we don't need to make strong assumptions about the users' use-case and instead we can provide a warning and a valid config, while still leaving the rest up to them.

Reply:
That _seems_ fair to me. I don't have much context here and can't really reason about the interactions of this modular sync and the data manager schedule sync, but I'm sure @AaronCasas will know more.

Reply:
Definitely think it's valid, just wondering if it's additional effort to support correctly, and if that additional work is necessary for the currently envisioned use case. I think thats a product question

Re this behavior being undetermined - in what ways is it undetermined right now? Would it not work how you describe (it syncing on a schedule, but maybe earlier if the trigger is hit)?

Reply:
I don't think undetermined is the right word as much as it might possibly be undesired behavior. For example, if sync is enabled and the sync interval is greater than the trigger polling interval, it may very well be that the selective sync never even happens. Open to any wording suggestions!